polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -152,105 +153,109 @@ async fn run<Context>(
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_chain_state();
-							let res = validate_from_chain_state(
-								&mut sender,
-								validation_host,
-								candidate_receipt,
-								pov,
-								executor_params,
-								timeout,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-chain-state", bg.boxed())?;
-				},
-				CandidateValidationMessage::ValidateFromExhaustive(
-					persisted_validation_data,
-					validation_code,
-					candidate_receipt,
-					pov,
-					executor_params,
-					timeout,
-					response_sender,
-				) => {
-					let bg = {
-						let metrics = metrics.clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_exhaustive();
-							let res = validate_candidate_exhaustive(
-								validation_host,
-								persisted_validation_data,
-								validation_code,
-								candidate_receipt,
-								pov,
-								executor_params,
-								timeout,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-exhaustive", bg.boxed())?;
-				},
-				CandidateValidationMessage::PreCheck(
-					relay_parent,
-					validation_code_hash,
-					response_sender,
-				) => {
-					let bg = {
-						let mut sender = ctx.sender().clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let precheck_result = precheck_pvf(
-								&mut sender,
-								validation_host,
-								relay_parent,
-								validation_code_hash,
-							)
-							.await;
-
-							let _ = response_sender.send(precheck_result);
-						}
-					};
-
-					ctx.spawn("candidate-validation-pre-check", bg.boxed())?;
-				},
+	let mut res = Ok(());
+	let sender = ctx.sender().to_owned();
+
+	let read_stream = stream::poll_fn(|c| loop {
+		match ctx.recv().poll_unpin(c) {
+			Poll::Ready(Ok(FromOrchestra::Signal(OverseerSignal::Conclude))) =>
+				return Poll::Ready(None),
+			Poll::Pending => return Poll::Pending,
+			Poll::Ready(Ok(FromOrchestra::Signal(_))) => continue,
+			Poll::Ready(Ok(FromOrchestra::Communication { msg })) => return Poll::Ready(Some(msg)),
+			Poll::Ready(Err(e)) => {
+				res = Err(e);
+				return Poll::Ready(None)
 			},
 		}
+	});
+
+	read_stream
+		// NB: Cloning `sender` inside `async` block of `for_each_concurrent` renders the whole
+		// thing not `Send`, so we `zip` the message stream with the stream of `sender` clones here.
+		.zip(stream::repeat(sender))
+		// FIXME: The backlog size here is the same as in PVF host queues. It should either use a
+		// common constant, or another appropriate value should be chosen
+		.for_each_concurrent(30, |message_and_sender| {
```

Comment:
We can, but I'm not totally sure it makes sense. If the candidate validation messages were just retransmitted to the PVF host's queue, it would be fine. But some invariants (like `ValidateFromChainState`) require candidate validation subsys to do additional work involving runtime requests on which it `await`s. So, if 30 `ValidateFromChainState` messages arrive, they are processed concurrently, and the PVF host queue is empty. If some `ValidateFromExhaustive` arrives at the moment, it is waiting until a free slot appears. But it could indeed fall through into the queue as it doesn't require any additional processing. That's why I think the limit here could be higher than the queue limit.
On the other hand, 30 concurrent `ValidateFromChainState`s is an unlikely situation. The real message stream is mixed `Validate*` requests with a handful of pre-checking requests. So I don't know, I just cannot decide which value is appropriate here.